README.md
# connect-oriento

Orient-DB session store for [Connect](https://github.com/senchalabs/connect) and [Express](http://expressjs.com/)
based on the binary transfer protocol in `oriento`.

## Installation

```shell
npm install connect-oriento
```

Written and tested on `io.js`, but should contain no `io.js` specific code and be runnable on `node.js`.


## Usage

### Express or Connect integration

Express `4.x`, `5.0` and Connect `3.x`:

```js
var session = require('express-session'),
    OrientoStore = require('connect-oriento')(session);

// minimalistic config providing only DB data
// provide server config as object or url encoded
var config = {
        session: {
            server: "host=localhost&port=2424&username=dev&password=dev&db=test"
        }
    };

app.use(session({
        secret: 'SomeSecret',
        store: new OrientoStore(config.session)
    }));
```

The full set of configuration options for the `OrientoStore` included any options accepted by
`express.session.Store` or `connect.Store`, respectively, plus the following specific options,
with the following defaults:

```js
var ORIENT_DEFAULTS = {
        // 1s: the session will not be touched until expired since last touch
        minTouchInterval: 1000,
        // every 1m Orient-DB server will be pinged to keep transport layer alive
        pingInterval: 1000 * 60,
        // every 10m will be checked if expired sessions should be deleted
        purgeInterval: 1000 * 600,
        // the session will be purged from the DB after 10 days (since last touch/set)
        maxAge: 1000 * 3600 * 24 * 10, // 10 days
        // should SessionID values be hashed when used as keys in the DB
        // to activeate specify an object with `algorithm` and `salt`
        hash: false
    };
```

When the session cookie has an expiration date, `connect-oriento` will use it instead of the `maxAge` option
described above. Otherwise, it will create a new one, using the `maxAge` option.

## Tests

You need `mocha` and an active Orient-DB instance:

```shell
export TEST_ORIENTDB="host=localhost&port=2424&username=root&password=root&db=test"
npm test
```

## License

(The MIT License)

Copyright (c) 2015 Casey Banner &lt;osklyar@qucado.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


history.md
0.1.0 / 2015-05-01
==================

* Initial release


test/test_store.js
var storeLib = require("../src");

var session = require('express-session'),
    assert = require("assert"),
    async = require("async"),
    util = require("odm-util");

var Session = require("../src/session");


describe("Test OrientoStore", function() {

    if (!process.env.TEST_ORIENTDB) {
        it("needs TEST_ORIENTDB env var defined in order to test DBClass");
        
        return;
    }
    
    // Format: host=localhost&port=2424&username=root&password=root&db=test
    var OrientoStore = undefined,
        cleanupStore= undefined,
        options = {
            server: process.env.TEST_ORIENTDB
        };

    before(function(done) {
        OrientoStore = storeLib(session);
        cleanupStore = new OrientoStore(options, done);
    });

    after(function(done) {
        Session.drop(cleanupStore._db, done);
    });

    it("can get non-existent record as empty object", function(done) {
        var store = new OrientoStore(util.object.merge({}, options));
        store.get("122342345345", function(err, session) {
            assert(session == null);
            done(err);
        })
    });

    it("can set and get sessions unhashed", function(done) {
        var store = new OrientoStore(util.object.merge({}, options));
        var sid1 = "asfhq3nwbf", sid2 = "bfngq438owf";

        async.waterfall([
            function(done) {
                store.set(sid1, { cookie: {}, val: 25 }, done);
            },
            function(session, done) {
                assert(25 == session.val);
                /* update with a new value */
                store.set(sid1, { cookie: {}, val: 26 }, done);
            },
            function(session, done) {
                assert(26 == session.val);
                store.get(sid1, done);
            },
            function(session, done) {
                assert(26 == session.val);
                store.set(sid2, { cookie: {}, val: 34 }, done);
            },
            function(session, done) {
                assert(34 == session.val);
                /* update with a new value */
                store.set(sid2, { cookie: {}, val: 35 }, done);
            },
            function(session, done) {
                assert(35 == session.val);
                store.get(sid2, done);
            },
            function(session, done) {
                assert(35 == session.val);
                Session.find(store._db, "1=1 order by id", done);
            },
            function(items, done) {
                assert(2 == items.length);
                assert(sid1 == items[0].id);
                assert(new Date() >= items[0].touched);
                assert(sid2 == items[1].id);
                assert(new Date() >= items[1].touched);
                Session.delete(store._db, {}, done);
            }
        ], done);
    });

    it("can set and get sessions hashed", function(done) {
        var store = new OrientoStore(util.object.merge({ hash: { algorithm: "sha1", salt: "foo" } }, options));
        var sid1 = "asfhq3nwbf", sid2 = "bfngq438owf";
        var hSid1 = store._getHashedSid(sid1), hSid2 = store._getHashedSid(sid2);

        async.waterfall([
            function(done) {
                store.set(sid1, { cookie: {}, val: 25 }, done);
            },
            function(session, done) {
                assert(25 == session.val);
                /* update with a new value */
                store.set(sid1, { cookie: {}, val: 26 }, done);
            },
            function(session, done) {
                assert(26 == session.val);
                store.get(sid1, done);
            },
            function(session, done) {
                assert(26 == session.val);
                store.set(sid2, { cookie: {}, val: 34 }, done);
            },
            function(session, done) {
                assert(34 == session.val);
                /* update with a new value */
                store.set(sid2, { cookie: {}, val: 35 }, done);
            },
            function(session, done) {
                assert(35 == session.val);
                store.get(sid2, done);
            },
            function(session, done) {
                assert(35 == session.val);
                Session.find(store._db, "1=1 order by session.id", done);
            },
            function(items, done) {
                assert(2 == items.length);
                assert(hSid1 == items[0].id);
                assert(new Date() >= items[0].touched);
                assert(hSid2 == items[1].id);
                assert(new Date() >= items[1].touched);
                Session.delete(store._db, {}, done);
            }
        ], done);
    });

    it("can will purge expired only", function(done) {
        var store = new OrientoStore(util.object.merge({}, options));
        var sid1 = "asdlgertw3s", sid2 = "b3458wzfw4rt3";

        async.waterfall([
            function(done) {
                store.set(sid1, { cookie: {}, val: sid1 }, done);
            },
            function(session, done) {
                store.set(sid2, { cookie: {}, val: sid2 }, done);
            },
            function(session, done) {
                Session.findById(store._db, sid2, done);
            },
            function(dbInstance, done) {
                dbInstance.expiry = new Date(Date.now() - 10000);
                dbInstance.save(store._db, done);
            },
            function(dbInstance, done) {
                store._purgeExpired(done);
            },
            function(totalPurged, done) {
                assert(1 == totalPurged);
                Session.find(store._db, done);
            },
            function(items, done) {
                assert(1 == items.length);
                assert(sid1 == items[0].id);
                Session.delete(store._db, {}, done);
            }
        ], done);
    });

    it("will purge on schedule", function(done) {
        var store = new OrientoStore(util.object.merge({ purgeInterval: 10 }, options));
        var sid1 = "asi347f8gf443", sid2 = "b34pwthalifg34";

        async.waterfall([
            function(done) {
                store.set(sid1, { cookie: {}, val: sid1 }, done);
            },
            function(session, done) {
                store.set(sid2, { cookie: {}, val: sid2 }, done);
            },
            function(session, done) {
                Session.findById(store._db, sid2, done);
            },
            function(dbInstance, done) {
                dbInstance.expiry = new Date(Date.now() - 10000);
                dbInstance.save(store._db, done);
            },
            function(dbInstance, done) {
                setTimeout(done, 25); // make sure we go over 1 purge iteration
            },
            function(done) {
                Session.find(store._db, done);
            },
            function(items, done) {
                assert(1 == items.length);
                assert(sid1 == items[0].id);
                Session.delete(store._db, {}, done);
            }
        ], done);
    });

    it("can destroy sessions", function(done) {
        var store = new OrientoStore(util.object.merge({}, options));
        var sid1 = "ae5tzsergh54", sid2 = "bw4tqw54z35wr";

        async.waterfall([
            function(done) {
                store.set(sid1, { cookie: {}, val: sid1 }, done);
            },
            function(session, done) {
                store.set(sid2, { cookie: {}, val: sid2 }, done);
            },
            function(session, done) {
                store.destroy(sid1, done);
            },
            function(count, done) {
                assert(1 == count);
                Session.find(store._db, done);
            },
            function(items, done) {
                assert(1 == items.length);
                assert(sid2 == items[0].id);
                Session.delete(store._db, {}, done);
            }
        ], done);
    });
    
    it("will touch only after interval", function(done) {
        var store = new OrientoStore(util.object.merge({ minTouchInterval: 50 }, options));
        var sid1 = "a45zseh5w45z";

        async.waterfall([
            function(done) {
                store.set(sid1, { cookie: {}, val: 25 }, done);
            },
            function(session, done) {
                setTimeout(done, 20);
            },
            function(done) {
                store.touch(sid1, {}, done);
            },
            function(when, done) {
                assert(19 < new Date() - when); // touch did not update
                setTimeout(done, 30);
            },
            function(done) {
                store.touch(sid1, {}, done);
            },
            function(when, done) {
                assert(40 > new Date() - when); // touched
                Session.delete(store._db, {}, done);
            }
        ], done);
    });

    it("can get the number of stored sessions", function(done) {
        var store = new OrientoStore(util.object.merge({}, options));
        var sid1 = "aertgaselbf348", sid2 = "bqw4ptrhgqfg";

        async.waterfall([
            function(done) {
                store.set(sid1, { cookie: {}, val: sid1 }, done);
            },
            function(session, done) {
                store.set(sid2, { cookie: {}, val: sid2 }, done);
            },
            function(session, done) {
                store.length(done);
            },
            function(count, done) {
                assert(2 == count);
                Session.delete(store._db, {}, done);
            }
        ], done);
    });

    it("can clear all sessions from the DB", function(done) {
        var store = new OrientoStore(util.object.merge({}, options));
        var sid1 = "asezegtg5seg", sid2 = "bw34twehgrzh";

        async.waterfall([
            function(done) {
                store.set(sid1, { cookie: {}, val: sid1 }, done);
            },
            function(session, done) {
                store.set(sid2, { cookie: {}, val: sid2 }, done);
            },
            function(session, done) {
                store.clear(done);
            },
            function(count, done) {
                assert(2 == count);
                store.length(done);
            },
            function(count, done) {
                assert(0 == count);
                done();
            }
        ], done);
    });

});


src/index.js
var util = require('util'),
    oriento = require("oriento");

var defaults = require("./defaults"),
    DISCONNECTED = -1,
    // CONNECTING = 0,
    CONNECTED = 1;


var initForSession = function(connect) {
    var Store = connect.Store || connect.session.Store;

    function OrientoStore(options, callback) {
        var self = this;
        try {
            /* "super" constructor */
            Store.call(this, defaults.ensureGenericStoreOptions(options));
            self._options = defaults.ensureOrientoStoreOptions(options);
        } catch(err) {
            if (callback) {
                return callback(err);
            } else {
                throw err;
            }
        }

        self._state = DISCONNECTED;
        self._ensureConnection(callback);
    }

    OrientoStore.prototype.get = function(sid, callback) {
    };

    OrientoStore.prototype.set = function(sid, session, callback) {
    };

    OrientoStore.prototype.touch = function(sid, session, callback) {
    };

    OrientoStore.prototype.destroy = function(sid, callback) {
    };

    OrientoStore.prototype.length = function(callback) {
    };

    OrientoStore.prototype.clear = function(callback) {
    };

    OrientoStore.prototype._ensureConnection = function(callback) {
        var self = this;
        if (!callback) {
            callback = defaults.returnOneCallback;
        }
        if (self._state == CONNECTED && self._db) {
            return callback(null, self._db);
        }
        var err, db;
        try {
            self._db = oriento(self._options.server).use(self._options.db);
            self._state = CONNECTED;
            self._scheduleDbPing();
        } catch(ex) {
            err = ex;
            self._state = DISCONNECTED;
        }
        return callback(err, self._db);
    };

    OrientoStore.prototype._scheduleDbPing = function() {
        var interval = this._options.pingInterval;
        /* init DB ping to now */
        this._db.ping = new Date();
        setInterval(pingDb, interval, this._db, interval);
    };

    util.inherits(OrientoStore, Store);
    return OrientoStore;
};


module.exports = initForSession;


/**
 * @deprecated Oriento drops the binary connection, so keep this one until the issue is resolved
 */
var pingDb = function(db, interval) {
    var now = new Date();
    // do not ping the same DB multiple times
    if (now - db.ping >= interval) {
        db.ping = now;
        db.query("SELECT FROM OUser").then(function() {
        }).catch(function(err) {
            console.error(err);
        });
    }
};


src/store.js
var async = require("async"),
    crypto = require("crypto"),
    util = require("odm-util"),
    oriento = require("oriento");

var Session = require("./session"),
    defaults = require("./defaults");
    
var INIT = -2,
    DISCONNECTED = -1,
    CONNECTING = 0,
    CONNECTED = 1;


var initForSession = function(connect, callback) {
    var Store = connect.Store || connect.session.Store;

    function OrientoStore(options, callback) {
        try {
            Store.call(this, defaults.ensureGenericStoreOptions(options));
    
            this._options = defaults.ensureOrientoStoreOptions(options);
            this._setState(INIT);
            this._ensureConnection();
        } catch(err) {
            if (callback) {
                callback(err);
                return;
            } else {
                throw err;
            }
        }
        
        /* ensure DB has the class and the index on id */
        this._ensureDBClass(callback);
        
        /* setup purge scheduler */
        setInterval(function(store) { store._purgeExpired(); }, this._options.purgeInterval, this);

        /* ugly Oriento issue of connection timeout */
        var interval = this._options.pingInterval;
        /* init DB ping to now */
        this._db.ping = new Date();
        setInterval(pingDb, interval, this._db, interval);
    }

    util.clazz.inherit(OrientoStore, Store);

    OrientoStore.prototype.get = function(sid, callback) {
        var self = this, db = self._db;
        sid = self._getHashedSid(sid);
        return async.waterfall([
            function(done) {
                done(self._state < CONNECTED ? new Error("disconnected from the DB") : null);
            },
            function(done) {
                Session.findById(db, sid, function(err, dbInstance) {
                    var session = null;
                    if (!err && dbInstance) {
                        session = JSON.parse(dbInstance.session || "{}");
                        if (!session.cookie) {
                            session = null;
                            err = new Error("persisted session is expected to have a cookie");
                        }
                    }
                    done(err, session);
                });
            }
        ], callback || defaults.returnOneCallback);
    };

    OrientoStore.prototype.set = function(sid, session, callback) {
        var self = this, db = self._db;
        sid = self._getHashedSid(sid);
        return async.waterfall([
            function(done) {
                done(self._state < CONNECTED ? new Error("disconnected from the DB") : null);
            },
            function(done) {
                done(!(session && session.cookie) ? new Error("session is expected to have a cookie") : null);
            },
            function(done) {
                Session.findById(db, sid, done);
            },
            function(dbInstance, done) {
                if (!dbInstance) {
                    // new session, otherwise update
                    dbInstance = new Session({ id: sid });
                }
                dbInstance.session = JSON.stringify(session);
                dbInstance.touched = new Date();
                dbInstance.expiry = self._computeExpiry(session);
                dbInstance.save(db, function(err, dbInstance) {
                    var session = {};
                    if (!err && dbInstance) {
                        session = JSON.parse(dbInstance.session || "{}");
                        if (!session.cookie) {
                            session = null;
                            err = new Error("persisted session is expected to have a cookie");
                        }
                    }
                    done(err, session);
                });
            }
        ], callback || defaults.returnOneCallback);
    };

    OrientoStore.prototype.touch = function(sid, session, callback) {
        var self = this, db = self._db;
        sid = self._getHashedSid(sid);
        return async.waterfall([
            function(done) {
                done(self._state < CONNECTED ? new Error("disconnected from the DB") : null);
            },
            function(done) {
                Session.findById(db, sid, done);
            },
            function(dbInstance, done) {
                done(!(dbInstance && dbInstance.session) ? new Error("can only touch already persisted session") : null, dbInstance);
            },
            function(dbInstance, done) {
                var now = new Date();
                if (now - dbInstance.touched < self._options.minTouchInterval) {
                    // do not touch if interval too small
                    return done(null, dbInstance.touched);
                }
                dbInstance.touched = now;
                dbInstance.expiry = self._computeExpiry(session);
                dbInstance.save(db, function(err) {
                    done(err, now);
                });
            }
        ], callback || defaults.returnOneCallback);
    };

    OrientoStore.prototype.destroy = function(sid, callback) {
        var self = this, db = self._db;
        sid = self._getHashedSid(sid);
        return async.waterfall([
            function(done) {
                done(self._state < CONNECTED ? new Error("disconnected from the DB") : null);
            },
            function(done) {
                Session.delete(db, { id: sid }, done);
            }
        ], callback || defaults.returnOneCallback);
    };

    OrientoStore.prototype.length = function(callback) {
        var self = this, db = self._db;
        return async.waterfall([
            function(done) {
                done(self._state < CONNECTED ? new Error("disconnected from the DB") : null);
            },
            function(done) {
                db.select("count(*)").from(Session.name).scalar().then(function(total) {
                    done(null, total);
                }).catch(done);
            }
        ], callback || defaults.returnOneCallback);
    };

    OrientoStore.prototype.clear = function(callback) {
        var self = this, db = self._db;
        return async.waterfall([
            function(done) {
                done(self._state < CONNECTED ? new Error("disconnected from the DB") : null);
            },
            function(done) {
                Session.delete(db, {}, done);
            }
        ], callback || defaults.returnOneCallback);
    };

    OrientoStore.prototype._ensureConnection = function(callback) {
        var self = this;
        if (!callback) {
            callback = defaults.returnOneCallback;
        }
        if (self._state == CONNECTED && self._db) {
            return callback(null, self._db);
        }
        /* ignore that we may be already connecting */
        self._setState(CONNECTING);
        var err;
        try {
            self._db = oriento(self._options.server).use(self._options.server.db);
            self._setState(CONNECTED);
        } catch(ex) {
            err = ex;
            self._setState(DISCONNECTED);
        }
        return callback(err, self._db);
    };

    OrientoStore.prototype._ensureDBClass = function(callback) {
        var self = this, db = self._db,
            idIndexName = Session.name + ".id";
        return async.waterfall([
            function(done) {
                done(self._state < CONNECTED ? new Error("disconnected from the DB") : null);
            },
            function(done) {
                db.class.get(Session.name).then(function(DBSessionClass) {
                    done(null, DBSessionClass);
                }).catch(function() {
                    db.class.create(Session.name, "V").then(function(DBSessionClass) {
                        console.log("created " + Session.name + " cluster");
                        done(null, DBSessionClass);
                    }).catch(done);
                });
            },
            function(DBSessionClass, done) {
                DBSessionClass.property.get("id").then(function(prop) {
                    if (prop == null) {
                        DBSessionClass.property.create({ name: 'id', type: 'String' }).then(function() {
                            console.log("created " + idIndexName + " property");
                            done(null, DBSessionClass);
                        }).catch(done);
                    } else {
                        done(null, DBSessionClass);
                    }
                });
            },
            function(DBSessionClass, done) {
                db.index.get(idIndexName).then(function(){
                    done(null, DBSessionClass);
                }).catch(function() {
                    db.index.create({ name: idIndexName, type: 'unique'}).then(function() {
                        console.log("created " + idIndexName + " index");
                        done(null);
                    }).catch(done);
                });
            }
        ], callback || defaults.returnOneCallback);
    };

    OrientoStore.prototype._getHashedSid = function(sid) {
        var hash = this._options.hash;
        if (hash) {
            return crypto.createHash(hash.algorithm).update(hash.salt + sid).digest('hex');
        } else {
            return sid;
        }
    };

    OrientoStore.prototype._setState = function(state) {
        var event;
        switch (state) {
            case DISCONNECTED:
                event = "disconnected";
                break;
            case CONNECTING:
                event = "connecting";
                break;
            case CONNECTED:
                event = "connected";
                break;
            default:
                event = "init";
        }
        this.emit(event);
        this._state = state;
    };

    OrientoStore.prototype._computeExpiry = function(session) {
        if (session && session.cookie && session.cookie.expires) {
            return new Date(session.cookie.expires);
        }
        return new Date(Date.now() + this._options.maxAge);
    };

    OrientoStore.prototype._purgeExpired = function(callback) {
        var self = this, db = self._db;
        callback = callback || defaults.returnOneCallback;
        db.query("DELETE VERTEX WHERE expiry < :date AND @class=:clazz", {
            params: {
                date: new Date(),
                clazz: Session.name
            }
        }).then(function(total) {
            callback(null, total);
        }).catch(callback);
    };

    return OrientoStore;
};


module.exports = initForSession;


/**
 * @deprecated Oriento drops the binary connection, 
 * so keep this one until the issue is resolved
 */
var pingDb = function(db, interval) {
    var now = new Date();
    // do not ping the same DB multiple times
    if (now - db.ping >= interval) {
        db.ping = now;
        db.query("SELECT FROM OUser").then(function() {
        }).catch(function(err) {
            console.error(err);
        });
    }
};


src/defaults.js
var oriento = require("oriento"),
    qs = require("qs");

var ORIENT_DEFAULTS = {
        minTouchInterval: 1000,        // 1 second
        pingInterval: 1000 * 60,       // 1 minute
        purgeInterval: 1000 * 600,     // 10 minutes
        maxAge: 1000 * 3600 * 24 * 10, // 10 days
        hash: false
    };

exports.ensureGenericStoreOptions = function(options) {
    var incoming = options || {},
        res = {};
    for (var key in incoming) {
        if (incoming.hasOwnProperty(key) && !ORIENT_DEFAULTS[key] && key != "server") {
            res[key] = incoming[key];
        }
    }
    return res;
};

exports.ensureOrientoStoreOptions = function(options) {

    var incoming = options || {},
        res = {};
    for (var key in ORIENT_DEFAULTS) {
        res[key] = incoming[key] || ORIENT_DEFAULTS[key];
    }

    /* the following mandatory fields do not have defaults */
    if (incoming.server) {
        res.server = typeof incoming.server == "string" ? qs.parse(incoming.server) : incoming.server;
    } else {
        throw new Error("'server' configuration missing");
    }

    if (res.hash && !(res.hash.algorithm && res.hash.salt)) {
        throw new Error("supply 'algorithm' and 'code' to enable SID hashing");
    }

    /* test connection */
    oriento(res.server).use(res.server.db);

    return res;
};

exports.returnOneCallback = function(err, data) {
    if (err) {
        throw err;
    }
    return data;
};
